internal/entimport/import.go

```diff
@@ -203,3 +205,157 @@ func isJoinTable(table *schema.Table) bool {
 	}
 	return true
 }
+
+func typeName(tableName string) string {
+	return inflect.Camelize(inflect.Singularize(tableName))
+}
+
+func tableName(typeName string) string {
+	return inflect.Underscore(inflect.Pluralize(typeName))
+}
+
+func resolvePrimaryKey(importer SchemaImporter, table *schema.Table) (f ent.Field, err error) {
+	if table.PrimaryKey == nil || len(table.PrimaryKey.Parts) != 1 {
+		return nil, fmt.Errorf("entimport: invalid primary key - single part key must be present")
+	}
+	if f, err = importer.field(table.PrimaryKey.Parts[0].C); err != nil {
+		return nil, err
+	}
+	if f.Descriptor().Name != "id" {
+		f.Descriptor().StorageKey = f.Descriptor().Name
+		f.Descriptor().Name = "id"
+	}
+	return f, nil
+}
+
+func upsertNode(i SchemaImporter, table *schema.Table) (*schemast.UpsertSchema, error) {
+	upsert := &schemast.UpsertSchema{
+		Name: typeName(table.Name),
+	}
+	fields := make(map[string]ent.Field, len(upsert.Fields))
+	for _, f := range upsert.Fields {
+		fields[f.Descriptor().Name] = f
+	}
+	pk, err := resolvePrimaryKey(i, table)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := fields[pk.Descriptor().Name]; !ok {
+		fields[pk.Descriptor().Name] = pk
+		upsert.Fields = append(upsert.Fields, pk)
+	}
+	for _, column := range table.Columns {
+		if column.Name == table.PrimaryKey.Parts[0].C.Name {
+			continue
+		}
+		fld, err := i.field(column)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := fields[column.Name]; !ok {
+			fields[column.Name] = fld
+			upsert.Fields = append(upsert.Fields, fld)
+		}
+	}
+	for _, index := range table.Indexes {
+		if index.Unique && len(index.Parts) == 1 {
+			fields[index.Parts[0].C.Name].Descriptor().Unique = true
+		}
+	}
+	for _, fk := range table.ForeignKeys {
+		for _, column := range fk.Columns {
+			// FK / Reference column
+			fields[column.Name].Descriptor().Optional = true
+		}
+	}
+	return upsert, err
+}
+
+func applyColumnAttributes(f ent.Field, col *schema.Column) {
+	desc := f.Descriptor()
+	desc.Optional = col.Type.Null
+	for _, attr := range col.Attrs {
+		if a, ok := attr.(*schema.Comment); ok {
+			desc.Comment = a.Text
+		}
+	}
+}
+
+func schemaMutations(importer SchemaImporter, tables []*schema.Table) ([]schemast.Mutator, error) {
+	mutations := make(map[string]schemast.Mutator, len(tables))
```

Comment:
```suggestion
	mutations := make(map[string]schemast.Mutator)
```
`mutations` is possibly larger than its items, since there is a `continue` statement in the loop.

Reply:
I'm aware of that, but isn't it more efficient to allocate the memory in advance (even if it's a little more) than re-allocate when the map needs to grow?

Reply:
It is more performant in case of CPU cycles and less context switches but not in terms of memory ;-) 

Since this `entimport` is not a performance critical appication I prefer the more "clean" way. But that is personal preference I guess :-)